File: language/benchmarks/tokens/algorithms/quick-sort/quick-sort.ts
function greaterOrEqual(pivot: number, xs: number[]): number[] {
  return xs.filter((value) => value >= pivot);
}

function lessThan(pivot: number, xs: number[]): number[] {
  return xs.filter((value) => value < pivot);
}

function quickSort(xs: number[]): number[] {
  return quickSortInto([], xs);
}

function quickSortInto(acc: number[], xs: number[]): number[] {
  if (xs.length === 0) return acc;
  const [pivot, ...tail] = xs;
  return quickSortInto([pivot, ...quickSortInto(acc, greaterOrEqual(pivot, tail))], lessThan(pivot, tail));
}

File: language/benchmarks/tokens/algorithms/quick-sort/quick-sort.py
def greater_or_equal(pivot: int, xs: list[int]) -> list[int]:
    return [value for value in xs if value >= pivot]


def less_than(pivot: int, xs: list[int]) -> list[int]:
    return [value for value in xs if value < pivot]


def quick_sort(xs: list[int]) -> list[int]:
    return quick_sort_into([], xs)


def quick_sort_into(acc: list[int], xs: list[int]) -> list[int]:
    if len(xs) == 0:
        return acc
    pivot, *tail = xs
    return quick_sort_into([pivot] + quick_sort_into(acc, greater_or_equal(pivot, tail)), less_than(pivot, tail))
